File: src/dataset/cpp/collatz_conjecture_Exercism_thefullarcticfox/gemini/GeminiFlash_collatz_conjecture_v2.cpp
#include "collatz_conjecture.h"

namespace collatz_conjecture {
	int steps(int n) {
		if (n <= 0) {
			throw std::domain_error("Input must be a positive integer.");
		}

		int res = 0;
		unsigned int num = static_cast<unsigned int>(n); // Use unsigned int to avoid potential overflow

		while (num != 1) {
			if ((num & 1) == 0) { // Use bitwise AND for even number check
				num >>= 1;       // Use right bit shift for division by 2
			} else {
				if (num > (UINT_MAX - 1) / 3) {
					throw std::overflow_error("Potential overflow detected");
				}
				num = num * 3 + 1;
			}
			++res;
		}
		return res;
	}
} // namespace collatz_conjecture

File: src/dataset/go/collatz_conjecture_Exercism_go_thinkverse/gemini/GeminiFlash_collatz_conjecture_v2.go
package collatzconjecture

import "errors"

func CollatzConjecture(n int) (int, error) {
	if n <= 0 {
		return 0, errors.New("n must be a positive integer")
	}

	steps := 0
	for n != 1 {
		if n&1 == 0 {
			n >>= 1
		} else {
			n = n*3 + 1
		}
		steps++
	}

	return steps, nil
}